### PR TITLE
Allow auto approval of claims for pull payments

### LINKED ISSUE #1780: Pull Payments- Support auto approval of payouts
**Is your feature request related to a problem? Please describe.**
You currently need to manually approve payouts. When we add automated payouts, we need an option to have payouts automatically approved in order to be sent.

**Describe the solution you'd like**
Just a checkbox when creating the pull payment to enable this.

### PR DESCRIPTION
**Pull request by Kukks**
closes #1780

### CODE REVIEW

**Comment by britttttk** [comment]
- It gives the awaiting approval notification even if the box is checked. 
- On the send page, confirmed payouts using auto-approve do not have the payout amount.

**Comment by Kukks** [comment]
> * It gives the awaiting approval notification even if the box is checked.
> * On the send page, confirmed payouts using auto-approve do not have the payout amount.
Thanks for that, fixed :)

File: BTCPayServer/HostedServices/PullPaymentHostedService.cs [hunk 1]
```diff
@@ -352,13 +354,32 @@ private async Task HandleCreatePayout(PayoutRequest req)
                 try
                 {
                     await ctx.SaveChangesAsync();
+                    if (ppBlob.AutoApproveClaims)
```
Comment (NicolasDorier) [inline BTCPayServer/HostedServices/PullPaymentHostedService.cs:357]: really hate this hack. I would prefer you call the approve inside the `Claim` method. You can pass the AutoApproveClaims to the `ClaimsResponse`.

Comment (Kukks) [reply BTCPayServer/HostedServices/PullPaymentHostedService.cs:357]: Not sure what you mean. I tried a few other ways before this but there was too much logic at play that could go wrong if I tried to optimize..

**Comment by britttttk** [comment]
tACK - thanks for fixes @Kukks 

If anyone wants to see it:

image

**Comment by Kukks** [comment]
@NicolasDorier  can we get this merged so we can move on with pull payments? I'm not sure what you wanted me to change code-wise.

**Comment by NicolasDorier** [comment]
What bother me in this PR is that users does not know what it mean to "approve a payout".

It actually mean, in this context, to lockdown the rate as soon as the payout is requested.
Merging it as is is a very confusing UX.

**Comment by Kukks** [comment]
> What bother me in this PR is that users does not know what it mean to "approve a payout".
> 
> It actually mean, in this context, to lock the exchange rate as soon as the payout is requested. But why would it be useful for users?
> Merging it as is is a very confusing UX.
> It does not mean "signing automatically".
> Ping @pavlenex , I think will agree with me.
> 
> I think it is doing something:
> 
> 1. That is not useful for anybody
> 2. Will confuse everybody as it is not doing what a reasonable user would expect

I really disagree with all of that.  
Approving a payout means to me: that it is ready to be paid.  Locking the rate is a side-effect not the feature.

When we add transaction batching (either via payjoin or other ways like a timer or criteria), approving a payout would queue it to the next batch, where a transaction is created, signed and broadcasted.

**Comment by NicolasDorier** [comment]
> Locking the rate is a side-effect not the feature.

A side effect which is not intended, and actually the only effect that "approving" does at the moment.

**Comment by Kukks** [comment]
> > Locking the rate is a side-effect not the feature.
> 
> A side effect which is not intended, and actually the only effect that "approving" does at the moment.

Rome wasn't built in a day, there are many building blocks needed to make pull payments useful. This is just the first one.

**Comment by Kukks** [comment]
> > Locking the rate is a side-effect not the feature.
> 
> A side effect which is not intended, and actually the only effect that "approving" does at the moment.

Actually it is intended, when creating the pull payment, you've already decided to either lock the rate or allow the rate to be decided on the time the user creates the payout request.

**Comment by NicolasDorier** [comment]
@Kukks the only reason why there is an approval step before paying is for locking the rate. There is no other reason.
Then it should not be called "Auto approval" but "Lock the rate as soon as the payout is created."

**Comment by Kukks** [comment]
> @Kukks the only reason why there is an approval step before paying is for locking the rate. There is no other reason.
> Then it should not be called "Auto approval" but "Lock the rate as soon as the payout is created."

Meh, the state is `AwaitingApproval`, `automatic approval` fits. Locking rates only appliesto pull payments where the pull payment currenc and the payout options currencies are not the same. If you use the approval step for other things, it's completely different. One example is using it as a way to vet the destination.

**Comment by britttttk** [comment]
I'm not commenting on the exchange rate lock, but please let me know if we need to update [this warning in the docs](https://docs.btcpayserver.org/PullPayments/#additional-use-cases-for-the-pull-payments-feature) from the outcome of this.

For the UX, I agree @NicolasDorier auto approve sounds like it should auto-pay (send) from the merchant's perspective. 

However I also agree with @Kukks that auto-approval is useful to communicate.

For anyone who wants to also give feedback this is what it looks like: 

**Without auto-approval:**
image

From the merchant perspective, there is no way to approve it. Paying it is the only way to approve it. 

**With auto-approval**
image

image

They appear the same after payout from both the requester and payer after payment. 

I think that the the communication of awaiting approval and auto approval is a useful thing to communicate (waiting for work to be sent, etc.) and the ux is clear.

**Comment by Kukks** [comment]
> For the UX, I agree @NicolasDorier auto approve sounds like it should auto-pay (send) from the merchant's perspective.

This is planned and the main end-goal for this PR for me

**Comment by Kukks** [comment]
> They appear the same after payout from both the requester and payer after payment.

This is different in #1834

**Comment by MaiHistoryCompanyLimited** [comment]
{#1851}

**Pull request merged**